File: server/cells_interlinked/pipeline/decoding_modes.py
"""Decoding-mode definitions: which captured positions get NLA-decoded.

The AV is a one-vector-in-one-paragraph-out instrument. We can't change
that shape without retraining. What we CAN choose is which positions to
feed it. Faster modes trade granularity for wall-clock — useful when you
want a rough cross-run signal at scale rather than a per-token narrative.

per-token   : every output position (slowest, fullest signal)
every-3rd   : stride 3 (~3x faster, smooth narrative)
every-5th   : stride 5 (~5x faster, recommended for batch)
key-points  : 5 positions: first, ~25%, ~50%, ~75%, last (very fast)
"""

from __future__ import annotations

from typing import Literal

DecodingMode = Literal["per-token", "every-3rd", "every-5th", "key-points"]

DECODING_MODES: tuple[DecodingMode, ...] = (
    "per-token",
    "every-3rd",
    "every-5th",
    "key-points",
)


def select_indices(n: int, mode: str) -> list[int]:
    """Return sorted output positions to NLA-decode for a captured run of length n."""
    if n <= 0:
        return []
    if mode == "per-token":
        return list(range(n))
    if mode == "every-3rd":
        return list(range(0, n, 3))
    if mode == "every-5th":
        return list(range(0, n, 5))
    if mode == "key-points":
        if n <= 5:
            return list(range(n))
        return sorted({0, n // 4, n // 2, (3 * n) // 4, n - 1})
    raise ValueError(f"unknown decoding_mode: {mode!r}")


def normalize_mode(mode: str | None) -> DecodingMode:
    if mode is None or mode == "":
        return "per-token"
    if mode not in DECODING_MODES:
        raise ValueError(
            f"unknown decoding_mode {mode!r}; valid: {DECODING_MODES}"
        )
    return mode  # type: ignore[return-value]

File: web/lib/decodingModes.ts
export type DecodingMode =
  | "per-token"
  | "every-3rd"
  | "every-5th"
  | "key-points";

export const DECODING_MODES: DecodingMode[] = [
  "per-token",
  "every-3rd",
  "every-5th",
  "key-points",
];

export const DECODING_MODE_LABELS: Record<DecodingMode, string> = {
  "per-token": "PER-TOKEN",
  "every-3rd": "EVERY 3RD",
  "every-5th": "EVERY 5TH",
  "key-points": "KEY POINTS",
};

export const DECODING_MODE_DESCRIPTIONS: Record<DecodingMode, string> = {
  "per-token":
    "Per-token — every output position is NLA-decoded. Slowest, fullest signal. ~10s per position; an 80-token answer takes ~13 min.",
  "every-3rd":
    "Every 3rd — captures the narrative shape at one third of the cost. Smooth read; loses moment-to-moment detail.",
  "every-5th":
    "Every 5th — recommended for batch runs. ~5× faster than per-token while keeping enough positions to see drift across an answer.",
  "key-points":
    "Key points — five strategic positions only (first, ~25%, ~50%, ~75%, last). ~30-60s per probe. Coarse but very fast.",
};

export function isDecodingMode(s: string): s is DecodingMode {
  return (DECODING_MODES as string[]).includes(s);
}
